<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">

<!--  Version: Multiflex-5.4 / About                     -->
<!--  Type:    Design with sidebar                          -->
<!--  Date:    March 13, 2008                               -->
<!--  Design:  www.1234.info                                -->
<!--  License: Fully open source without restrictions.      -->
<!--           Please keep footer credits with the words    -->
<!--           "Design by 1234.info". Thank you!            -->

<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  <meta http-equiv="cache-control" content="no-cache" />
  <meta http-equiv="expires" content="3600" />
  <meta name="revisit-after" content="2 days" />
  <meta name="robots" content="index,follow" />
  <meta name="publisher" content="Your publisher infos here" />
  <meta name="copyright" content="Your copyright infos here" />
  <meta name="author" content="Designed by www.1234.info / Modified: Your Name" />
  <meta name="distribution" content="global" />
  <meta name="description" content="Your container description here" />
  <meta name="keywords" content="Your keywords, keywords, keywords, here" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_reset.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_grid.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_content.css" />
  <link rel="icon" type="image/x-icon" href="./img/favicon.ico" />
  <title>eXpress &bull; Manual</title>
</head>

<!-- Global IE fix to avoid layout crash when single word size wider than column width -->
<!-- Following line MUST remain as a comment to have the proper effect -->
<!--[if IE]><style type="text/css"> body {word-wrap: break-word;}</style><![endif]-->

<body>
  <!-- CONTAINER FOR ENTIRE PAGE -->
  <div class="container">

    <!-- A. HEADER -->         
    <div class="corner-page-top"></div>        
    <div class="header">
      <div class="header-top">
        
        <!-- A.1 SITENAME -->    
        <div class="sitelogo">
        <ul>
            <li><a href="#"><img src="img/logo.png"/></a></li>
        </ul>
        </div>
        <div class="sitename">
          <h1><a href="index.html" title="Home">eXpress</a></h1> 
          &nbsp;
          <h2><i>Streaming</i> RNA-Seq Analysis</h2>
        </div>
    
        <!-- A.2 BUTTON NAVIGATION -->
        <div class="navbutton">
          <ul>
            <li><a href="http://www.berkeley.edu"><img src="img/berkeley_seal.gif"/></a></li>
          </ul>
        </div>
      </div>

      <!-- A.4 BREADCRUMB and SEARCHFORM -->
      <div class="header-bottom">		

        <!-- Search form -->    
<div class="searchform" id="cse-search-form" style="padding-top:4px; width:30%;">Loading</div>
<script src="http://www.google.com/jsapi" type="text/javascript"></script>
<script type="text/javascript"> 
  google.load('search', '1', {language : 'en'});
  google.setOnLoadCallback(function() {
    var customSearchControl = new google.search.CustomSearchControl('017113487431020499685:jdftct1emrw');
    customSearchControl.setResultSetSize(google.search.Search.FILTERED_CSE_RESULTSET);
    var options = new google.search.DrawOptions();
    options.enableSearchboxOnly("http://www.google.com/cse?cx=017113487431020499685:jdftct1emrw", null, true);
    customSearchControl.draw('cse-search-form', options);
  }, true);
</script>
<link rel="stylesheet" href="css/googlesearch.css" type="text/css" />

      </div>
    </div>      
    <div class="corner-page-bottom"></div>    
    
    <!-- B. NAVIGATION BAR -->
    <div class="corner-page-top"></div>        
    <div class="navbar">
	
      <!-- Navigation item -->
      <ul>
        <li><a href="index.html">Home</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="overview.html">About</a></li>
      </ul>            
       <!-- Navigation item -->
      <ul>
        <li><a href="tutorial.html">Getting Started</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="manual.html">Manual</a></li>
      </ul>
      <ul>
        <li><a href="faq.html">FAQ</a></li>
      </ul>
      </div>
  
    <!-- C. MAIN SECTION -->      
    <div class="main">
    	<h1 class="pagetitle">Manual</h1>

      <!-- C.1 CONTENT -->
      <div class="content">
                                
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
		
		  <h1 id="usage">Usage</h1>
		  <h2 id="prereq">Prerequisites</h2>
		  <p>eXpress runs on intel-based computers running Linux, Mac OS X, or Windows.  You can install pre-compiled binaries or build eXpress from the source code. If you wish to build eXpress yourself, you must have a C++ compiler installed (for example, <a href=http://developer.apple.com/tools/xcode>XCode</a> for Mac OS X, <a href=http://www.microsoft.com/visualstudio/en-us/products/2010-editions/visual-cpp-express>Visual C++ Express</a> for Windows 7) as well as <a href=http://www.cmake.org>CMake</a>, <a href=https://github.com/pezmaster31/bamtools>BamTools</a>, and the <a href="http://www.boost.org">Boost C++ libraries</a>.  See the <a href="tutorial.html#install">Installation</a> section on the Getting Started page for detailed instruction.</p> 
		  <h2 id="running">Running eXpress</h2>
       	  <p>Run <tt>eXpress</tt> from the command line as follows:
	      <ul><pre class="sc"><code>$ express [options]* &lt;target_seqs.fasta&gt; &lt;aligned_reads.(sam/bam)&gt;</code></pre></ul></p> 
	  
		  <p>The following is a detailed description of the options used to control eXpress:
		  <table> 
		  <tr><td WIDTH=30% VALIGN=top><b>Arguments:</b></td><td WIDTH=70% VALIGN=top></td></tr> 
		  
		  <tr><td VALIGN=top nowrap><tt>&lt;target_seqs.fasta&gt;</tt></td>
		  <td VALIGN=top>A file of target sequences in <a href=http://en.wikipedia.org/wiki/FASTA_format>multi-FASTA</a> format. See <a href="#fasta">Input Files</a> for more details.</td></tr> 
		  
		  <tr><td VALIGN=top nowrap><tt>&lt;aligned_reads.(sam/bam)&gt;</tt></td>
		  <td VALIGN=top>A file of RNA-Seq read alignments in the <a href="http://samtools.sourceforge.net"> 
		  SAM format</a>. See <a href="#sam">Input Files</a> for more details.</td></tr> 
		  
		  <tr><td VALIGN=top><b>General Options:</b></td><td VALIGN=top></td></tr>
		  
		  <tr><td VALIGN=top nowrap><tt>-h/--help</tt> </td>
		  <td VALIGN=top>Prints the help message and exits</td></tr> 
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>-o/--output-dir &lt;string&gt;</tt> 
		  </td><td VALIGN=top> 
		  Sets the name of the directory in which eXpress will write all of its 
		  output.  The default is "./".
		  </td></tr> 
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>-m/--frag-len-mean &lt;int&gt;</tt> 
		  </td><td VALIGN=top> 
		  Specifies the mean fragment length.  While the empirical distribution is estimated from paired-end reads on-the-fly, this value paramaterizes the prior distribution.  If only single-end reads are available, this prior distribution is also used to determine the effective length.  Default is 200.
		  </td></tr> 
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>-s/--frag-len-stddev &lt;int&gt;</tt> 
		  </td><td VALIGN=top> 
		  Specifies the fragment length standard deviation.  While the empirical distribution is estimated from paired-end reads on-the-fly, this value paramaterizes the prior distribution.  If only single-end reads are available, this prior distribution is also used to determine the effective length.  Default is 60.
		  </td></tr> 
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>--fr-stranded</tt> 
		  </td><td VALIGN=top> 
		  With this option, eXpress only accepts alignments where the first read is aligned to the forward target sequence and the second read is aligned to the reverse-complemented target sequence.  In directional sequencing, this is equivalent to second-strand only. By default, both orientations are allowed. Mutually exclusive with --rf-stranded.
		  </td></tr>
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>--rf-stranded</tt> 
		  </td><td VALIGN=top> 
		  With this option, eXpress only accepts alignments where the first read is aligned to the reverse-completemented target sequence and the second read is aligned to the forward target sequence.  In directional sequencing, this is equivalent to first-strand only. By default, both orientations are allowed.  Mutually exclusive with --fr-stranded.
		  </td></tr>
		  
		  <tr><td VALIGN=top nowrap> 
		  <tt>--calc-covar</tt> 
		  </td><td VALIGN=top> 
		  With this option, eXpress calculates the covariance between targets and outputs them for use in differential expreession analysis. This calculates requires slightly more time and memory.
		  </td></tr>
		  
  		  </table></p>
		
		          </div> 
        <div class="corner-content-1col-bottom"></div>                               
      
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
		<h1 id="input">Input Files</h1>
		<h2 id="fasta">Target Sequences (FASTA)</h2>
	  	<p>eXpress takes a <a href=http://en.wikipedia.org/wiki/FASTA_format>multi-FASTA</a> file of target sequences to measure the abundance of. In the case of RNA-Seq, these are the transcript sequences.  If
	  	the transcriptome of your organism is not
        annotated, you can generate this file from your sequencing
        reads using a <i>de novo</i> transcriptome assembler such as
        <a href=http://trinityrnaseq.sourceforge.net/>Trinity</a>, <a
        href=http://www.ebi.ac.uk/~zerbino/oases/>Oases</a>, or <a
        href=http://www.bcgsc.ca/platform/bioinfo/software/trans-abyss>Trans-ABySS</a>.
        If your organism has a reference genome you can assemble
        transcripts directly from mapped reads using <a
        href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>. If your genome is already annotated (in
        GTF/GFF), you can generate a multi-FASTA file using the <a href=http://genome.ucsc.edu/>UCSC Genome Browser</a> by uploading your annotation as a track and downloading the sequences under the "Tables" tab.</p>
	  	
	  	<h2 id="sam">Read Alignments (SAM/BAM)</h2>
	  	<p>eXpress also requires a file or piped stream of SAM or binary SAM (BAM) alignments as input.  The SAM alignments should be generated by mapping your sequencing reads to the target sequences specified in the multi-FASTA input file described above.  For more details on the SAM format, see the <a href="http://samtools.sourceforge.net/SAM1.pdf">specification</a>.  Many short read mappers including <a href="http://bowtie-bio.sourceforge.net">Bowtie</a>, <a href="http://bio-bwa.sourceforge.net/">BWA</a>, and <a href=http://maq.sourceforge.net/>MAQ</a> can produce output in this format. It is important that you allow many multi-mappings (preferably unlimited) in order to get the most accurate results.  See <a href="tutorial.html#example">Getting Started</a> for an example using Bowtie in both streaming and file input modes.</p> 
	  <p> 
	  The SAM file supplied to eXpress should be sorted by read id.  If you aligned your reads with Bowtie, your 
	  alignments will be properly sorted already.  If you used another tool, you
	  should ensure that they are properly sorted as eXpress will not verify this for you.  You can sort your SAM using the following command:
	<ol><pre class="sc"><code>sort -k 1 hits.sam > hits.sam.sorted</code></pre></ol></p>
        
                  </div> 
        <div class="corner-content-1col-bottom"></div>                               
      
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
        <h1 id="output">Output Files</h1>
        <h2 id="expr">Target Abundances (results.xprs)</h2>
        <p>This file is always output and contains the target abundances and other values calculated based on the input sequences and read alignments. The file has 10 tab-delimited columns, sorted by the bundle_id (column 1). The columns are defined as follows: 
        <!--"bundle_id transcript_id length eff_length bundle_frac est_counts est_counts_var fpkm --!>
      <table CELLSPACING=15> 
	  <tr><th class="top" scope="col" width=11%>Column Number</th><th class="top" scope="col" width=17%>Column Name</th><th class="top" scope="col" width=17%>Example</th><th class="top" scope="col" width=55%>Description</th></tr> 
	  <tr><th scope="row">1</th><td>bundle_id</td><td><tt>48</tt></td><td>ID of bundle the target belongs to.  A bundle is defined as the transitive closure of targets that share multi-mapping reads.</td></tr> 
 	  <tr><th scope="row">2</th><td>target_id</td><td><tt>NM_001211</tt></td><td>The ID given to the target in the input multi-FASTA file.</td></tr> 
   	  <tr><th scope="row">3</th><td>length</td><td><tt>3749</tt></td><td>The number of base pairs in the target sequence given in the input multi-FASTA file.</td></tr> 
	  <tr><th scope="row">4</th><td>eff_length</td><td><tt>3596.22</tt></td><td>The effective length of the target taking into account the fragment length distribution and fragment biases. This number is what the fragment counts are normalized by to calculate FPKM, not the true length.</td></tr> 
   	  <tr><th scope="row">5</th><td>bundle_frac</td><td><tt>0.455</tt></td><td>The fraction of fragment counts probabilistically assigned to this target out of the total number of fragments mapping to targets in the bundle.</td></tr> 
   	  <tr><th scope="row">6</th><td>tot_counts</td><td><tt></tt></td><td>The number of fragments mapping (uniquely or ambiguously) to this target.</td></tr> 
   	  <tr><th scope="row">7</th><td>uniq_counts</td><td><tt></tt></td><td>The number of fragments uniquely mapping to this target.</td></tr> 
	  <tr><th scope="row">8</th><td>est_counts</td><td><tt>5020.468890</tt></td><td>The estimated number of fragments generated from this target in the sequencing experiment.</td></tr> 
  	  <tr><th scope="row">9</th><td>est_counts_var</td><td><tt></tt></td><td>The estimated variance on the number of fragments generated from this target in the sequencing experiment due to uncertainty in assignment of multi-mapped reads.</td></tr> 
   	  <tr><th scope="row">10</th><td>fpkm</td><td><tt></tt></td><td>The estimated relative abundance of this target in the sample in units of <b>f</b>ragments <b>p</b>er <b>k</b>ilobase per <b>m</b>illion sequenced. This value is proportional to est_counts divided by eff_length.</td></tr> 
	  </table></p>
	  <p>See the <a href="overview.html#methods">Methods</a> for more details on how these values are calculated.</p>
	  
      <h2 id="covar">Count Variance-Covariance (varcov.xprs)</h2>
      <p>This file is produced only when the <tt>--calc-covars</tt> option flag is used as described <a href="#running">above</a>.  The file contains the estimated variances and covariances on the counts between pairs of targets that shared multi-mapped reads, primarily to be used in differential expression analysis.  Each target will be listed with its variance on exactly one line (the target_id will be repeated in the first two columns).  Each pair of targets with a nonzero covariance will be listed on exactly one line of the file.  If a pair of targets is not listed in the file, it has a covariance of 0.  The 3 tab-delimited columns of the file are defined as follows:
      <table CELLSPACING=15> 
	  <tr><th class="top" scope="col" width=11%>Column Number</th><th class="top" scope="col" width=17%>Column Name</th><th class="top" scope="col" width=17%>Example</th><th class="top" scope="col" width=55%>Description</th></tr> 
	  <tr><th scope="row">1</th><td>target_1_id</td><td><tt>NM_001211</tt></td><td>The ID of the first target in the covariance pair (ordering is arbitrary).</td></tr> 
 	  <tr><th scope="row">2</th><td>target_2_id</td><td><tt>NM_001212</tt></td><td>The ID of the second target in the covariance pair (ordering is arbitrary).  This will be the same as target_1_id if the entry is for a variance.</td></tr> 
   	  <tr><th scope="row">3</th><td>covar</td><td><tt></tt></td><td>The estimated variance for the single target (if target_1_id==target_2_id) or covariance between the two targets due to uncertatinty in the assignment of multi-mapped reads.</td></tr> 
	  </table>
	  <p>See the <a href="overview.html#methods">Methods</a> for more details on how these values are calculated.</p> 
	  
        </div> 
        <div class="corner-content-1col-bottom"></div>                               
	  </div>
      
       <!-- C.2 SUBCONTENT -->
      <div class="subcontent">
      
          <!-- SUBCONTENT CELL -->
        <div class="corner-subcontent-top"></div>                        
        <div class="subcontent-box">
          <h1 class="menu">Outline</h1> 
          <div class="sidemenu1">
          <!-- CONTENT CELL -->                
 		  <ul>
 		  	<li><a href="#usage">Usage</a></li>
			 <li><a href="#prereq">&rarr;Prerequisites</a></li>
			 <li><a href="#running">&rarr;Running eXpress</a></li>
			<li><a href="#input">Input Files</a></li>
			 <li><a href="#fasta">&rarr;Target Sequences (FASTA)</a></li>
			 <li><a href="#sam">&rarr;Read Alignments (SAM/BAM)</a></li>
			<li><a href="#output">Output Files</a></li>
			 <li><a href="#expr">&rarr;Target Abundances (results.xprs)</a></li>
			 <li><a href="#covar">&rarr;Variance-Covariance (varcov.xprs)</a></li>
		  </ul>
		  </div> 
         </div>
        <div class="corner-subcontent-bottom"></div> 

 	 </div>
    </div>
      
    <!-- D. FOOTER -->      
    <div class="footer">
      <p>Copyright &copy; 2011 Adam Roberts&nbsp;&nbsp;|&nbsp;&nbsp;All Rights Reserved</p>
      <p class="credits">Design by <a href="http://1234.info/" title="Designer Homepage">1234.info</a> | Modified by <a href="http://cs.berkeley.edu/~adarob/">Adam Roberts</a> | <a href="http://validator.w3.org/check?uri=referer" title="Validate XHTML code">XHTML 1.0</a> | <a href="http://jigsaw.w3.org/css-validator/" title="Validate CSS code">CSS 2.0</a></p>
      <br />
      <p>The eXpress project was funded in part by an NSF graduate fellowship to Adam Roberts and NIH grant 1R01HG006129-01</p>
    </div>
    <div class="corner-page-bottom"></div>        
  </div>
  
</body>
</html>



