<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">

<!--  Version: Multiflex-5.4 / About                     -->
<!--  Type:    Design with sidebar                          -->
<!--  Date:    March 13, 2008                               -->
<!--  Design:  www.1234.info                                -->
<!--  License: Fully open source without restrictions.      -->
<!--           Please keep footer credits with the words    -->
<!--           "Design by 1234.info". Thank you!            -->

<head>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-25486066-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  <meta http-equiv="cache-control" content="no-cache" />
  <meta http-equiv="expires" content="3600" />
  <meta name="revisit-after" content="2 days" />
  <meta name="robots" content="index,follow" />
  <meta name="publisher" content="Your publisher infos here" />
  <meta name="copyright" content="Copyright (c) 2011 Adam Roberts" />
  <meta name="author" content="Designed by www.1234.info / Modified: Adam Roberts" />
  <meta name="distribution" content="global" />
  <meta name="image" content="http://bio.math.berkeley.edu/eXpress/img/logo.png" />
  <meta name="description" content="eXpress is a general quantification tool for target DNA/RNA sequences. While its primary use currently is RNA-Seq it has the potential for applications in many other areas including as allele-specific expression and metgenomics. What makes eXpress different is that it is an online (or streaming) algorithm, meaning it only makes one pass through the data. This allows it to be very light-weight and efficient using a constant amount of memory and time linear in the number of sequenced fragments being processed. Furthermore, it accepts piped SAM/BAM input, allowing users to avoid storing extremely large alignment files. eXpress models fragment biases, fragment lengths, and errors, allowing it to also be one of the most accurate quantification methods available." />
  <meta name="keywords" content="RNA-Seq, Genomics, transcript,quantification" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_reset.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_grid.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_content.css" />
  <link rel="icon" type="image/x-icon" href="./img/favicon.ico" />
   <script language="Javascript">
        logo1=new Image
        logo1.src="img/logo.png"
        logo2=new Image
        logo2.src="img/logo_white.png"
 </script>
  <title>eXpress &bull; Getting Started</title>
</head>
<link rel="shortcut icon" href="favicon.ico" />

<!-- Global IE fix to avoid layout crash when single word size wider than column width -->
<!-- Following line MUST remain as a comment to have the proper effect -->
<!--[if IE]><style type="text/css"> body {word-wrap: break-word;}</style><![endif]-->



<body>
  <!-- CONTAINER FOR ENTIRE PAGE -->
  <div class="container">

    <!-- A. HEADER -->         
    <div class="corner-page-top"></div>        
    <div class="header">
      <div class="header-top">
        
        <!-- A.1 SITENAME -->    
        <div class="sitelogo">
        <ul>
            <li><a href="#" onMouseOver="document.logo.src=logo2.src" onMouseOut="document.logo.src=logo1.src"><img name="logo" src="img/logo.png"/></a></li>
        </ul>
        </div>
        <div class="sitename">
          <h1><a href="#">eXpress</a></h1> 
          &nbsp;
          <h2><i>Streaming</i> RNA-Seq Analysis</h2>
        </div>
    
        <!-- A.2 BUTTON NAVIGATION -->
        <div class="navbutton">
          <ul>
            <li><a href="http://www.berkeley.edu"><img src="img/berkeley_seal.gif"/></a></li>
          </ul>
        </div>
      </div>

      <!-- A.4 BREADCRUMB and SEARCHFORM -->
      <div class="header-bottom">

        <!-- Search form -->    
<div class="searchform" id="cse-search-form" style="padding-top:4px; width:30%;">Loading</div>
<script src="http://www.google.com/jsapi" type="text/javascript"></script>
<script type="text/javascript"> 
  google.load('search', '1', {language : 'en'});
  google.setOnLoadCallback(function() {
    var customSearchControl = new google.search.CustomSearchControl('017113487431020499685:jdftct1emrw');
    customSearchControl.setResultSetSize(google.search.Search.FILTERED_CSE_RESULTSET);
    var options = new google.search.DrawOptions();
    options.enableSearchboxOnly("http://www.google.com/cse?cx=017113487431020499685:jdftct1emrw", null, true);
    customSearchControl.draw('cse-search-form', options);
  }, true);
</script>
<link rel="stylesheet" href="css/googlesearch.css" type="text/css" />

   </div>
    </div>      
    <div class="corner-page-bottom"></div>    
    
    <!-- B. NAVIGATION BAR -->
    <div class="corner-page-top"></div>        
    <div class="navbar">
	
      <!-- Navigation item -->
      <ul>
        <li><a href="index.html">Home</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="overview.html">About</a></li>
      </ul>            
      <!-- Navigation item -->
      <ul>
        <li><a href="#">Download<!--[if IE 7]><!--></a><!--<![endif]-->
          <!--[if lte IE 6]><table><tr><td><![endif]-->
            <ul>
              <li><a href=downloads/express-0.9.5/express-0.9.5-macosx_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Mac', 'Tutorial']);" target="_blank">Mac OS X (64-bit)</a></li>
              <li><a href=downloads/express-0.9.5/express-0.9.5-linux_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Linux', 'Tutorial']);" target="_blank">Linux (64-bit)</a></li>
              <li><a href=downloads/express-0.9.5/express-0.9.5-win32_x86_64.zip onClick="_gaq.push(['_trackEvent', 'Downloads', 'Windows', 'Tutorial']);" target="_blank">Windows (64-bit)</a></li>
              <li><a href=downloads/express-0.9.5/express-0.9.5-src.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Source', 'Tutorial']);" target="_blank">Source</a></li>
              <li><a href="downloads" onClick="_gaq.push(['_trackEvent', 'Downloads', 'Previous','Tutorial']);">Previous Versions</a></li>
             </ul>
          <!--[if lte IE 6]></td></tr></table></a><![endif]-->
        </li>
      </ul>                  
       <!-- Navigation item -->
      <ul>
        <li><a href="tutorial.html">Getting Started</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="manual.html">Manual</a></li>
      </ul>
      <ul>
        <li><a href="faq.html">FAQ</a></li>
      </ul>
      </div>
  
    <!-- C. MAIN SECTION -->      
    <div class="main">
		<h1 class="pagetitle">Getting Started</h1>
      <!-- C.1 CONTENT -->
      <div class="content">
                                    
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">

          <h1 id="install">Installation</h1>
		  <h2 id="precomp">Installing a pre-compiled binary release</h2>
		  <p>In order to make it easy to install eXpress, we provide a few binary packages that save you the trouble of having to compile from source. To use the binary packages for OSX/Linux (Windows), simply download the appropriate one for your machine, untar (unzip) it, and make sure the <code>express(.exe)</code> binary is in a directory in your <tt>PATH</tt> environment variable.</p>
		  <p>If using the Windows binary, you will also need to install the <a href="http://www.microsoft.com/download/en/details.aspx?id=5555">Visual C++ 2010 Runtime Library</a>.</p>
		  <h2 id="src">Installing from source</h2>
		  <p><b>Note:</b> Installing from source is often an unnecessary hassle.  If a binary is available for your system, it is highly recommended that you use it.  If one is not available, please sendd us a request to <a href="mailto:ask.xprs@gmail.com">ask.xprs@gmail.com</a>.</p>
		  <p>The instructions below include commands that are compatible with Unix/Linux/Mac. If compiling on a Windows machine, use the appropriate DOS commands instead.  Assuming you are using the Visual C++ compiler, you must type your commands from the Visual Studio Command Prompt.
		  <h4>&bull; Install C++ Compiler (if not already available)</h4>
		  <p>While most Linux systems will already have GCC installed, Mac OS X and Windows do not have a C++ compiler installed by default.  Free options include <a href=http://developer.apple.com/tools/xcode>XCode</a> for Mac OS X and <a href=http://www.microsoft.com/visualstudio/en-us/products/2010-editions/visual-cpp-express>Visual C++ Express</a> for Windows.</p>
		  <h4>&bull; Download and extract the source</h4>
		  <p>First you must download source code using the link on the right side of this page.  Untar the file.  From now on we refer to the path to the directory that is created as <code>&lt;YOUR_EXPRESS_DIR&gt;</code>.
          <h4>&bull; Install CMake</h4>
          <p>Fortunately, the developers of CMake provide simple installation packages for most architectures. Find the right package for your system at their <a href=http://www.cmake.org/cmake/resources/software.html>website</a>.</p>
          <h4>&bull; Install BamTools</h4>
          <p>With CMake installed, BamTools installation is straightforward.</p>
          <ol>
          	<li>Download the source from <a href=https://github.com/pezmaster31/bamtools/tarball/master>here</a> and untar into <code>&lt;YOUR_EXPRESS_DIR&gt;</code>.</li>
          	<li>Navigate to <code>&lt;YOUR_EXPRESS_DIR&gt;/bamtools</code>.
          	<pre class="sc"><code>$ cd &lt;YOUR_EXPRESS_DIR&gt;/bamtools</code></pre></li>
          	<li>Make a new directory called build and navigate to it.
          	<pre class="sc"><code>$ mkdir build<br />$ cd build</code></pre></li>
          	<li>Have CMake generate the makefile.
          	<pre class="sc"><code>$ cmake ..</code></pre></li>
          	<li>Build the BamTools libraries by calling make.
          	<pre class="sc"><code>$ make</code></pre></li>
          </ol>
          <h4>&bull; Install BOOST</h4>
          <p>If you have Ports or MacPorts installed, these will probably be the easiest way to install BOOST.  If you are using Windows, you can use the installer found <a href=http://www.boostpro.com/download/>here</a>.  Otherwise, follow the instructions below.</p>
          			<ol>
                    <li><a href="http://www.boost.org/users/download/">Download</a>
                      Boost and the <code>bjam</code> build engine.</li>
                    <li>Unpack bjam and add it to your <tt>PATH</tt>.</li>
                    <li>Unpack the Boost tarball and <code>cd</code> to the
                      Boost source directory. This directory is called
                      the <code>BOOST_ROOT</code> in some Boost installation
                      instructions.</li>
                    <li>Build Boost. Note that you can specify where to
                      put Boost with the <code>--prefix</code> option. The default
                      Boost installation directory is <code>/usr/local</code>.
					</li></ol>
					<ul><ul>

                    <li>If you are on Mac OS X, type:<br />
                    <pre class="sc"><code>$ bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=darwin architecture=x86 address_model=32_64 link=static runtime-link=static --layout=versioned stage install</code></pre></li>
                    <li>If you are on a 32-bit Linux system, type:<br />
                    <pre class="sc"><code>$ bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=gcc architecture=x86 address_model=32 link=static runtime-link=static stage install</code></pre></li>
                      
                    <li>If you are on a 64-bit Linux system, type:<br />
                    <pre class="sc"><code>$ bjam --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt; --toolset=gcc architecture=x86 address_model=64 link=static runtime-link=static stage install</code></pre></li>
                    </ul> </ul>                    
      	<h4>&bull; Install eXpress</h4>
          <p>We are now ready to build and install eXpress!</p>
          <ol>
          	<li>Navigate to <code>&lt;YOUR_EXPRESS_DIR&gt;</code>.
          	<pre class="sc"><code>$ cd &lt;YOUR_EXPRESS_DIR&gt;</code></pre></li>
          	<li>Make a new directory called build and navigate to it.
          	<pre class="sc"><code>$ mkdir build<br />$ cd build</code></pre></li>
          	<li>Have CMake generate the makefile.
          	<pre class="sc"><code>$ cmake ..</code></pre></li>
          	<li>Build the eXpress binary by calling make.
          	<pre class="sc"><code>$ make</code></pre></li>
          	<li>Copy the binary to <tt>/usr/lib/bin</tt> (or alternatively to another directory in your <tt>PATH</tt>).
          	<pre class="sc"><code>$ sudo make
                    install</code></pre></li>
         </ol>
          <p>You should now be able to type <tt>express</tt> from any directory and see a print-out of the usage and options.  If you do not see this and there were no errors in the compilation, double check to see that the binary was copied into a directory in your <tt>PATH</tt>.</p>
          		<p>&rarr; <a href="#top">Back to top.</a></p>
          		 </div>
        <div class="corner-content-1col-bottom"></div>                               

        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
          <h1 id="usecase">General Use Case: RNA-Seq abundances</h1>
		  <h2 id="reqin">Required input</h2>
		  <p>eXpress requires two input files:</p>
		  <ol>
		  <li>A multi-FASTA file containing the transcript
        sequences. If the transcriptome of your organism is not
        annotated, you can generate this file from your sequencing
        reads using a <i>de novo</i> transcriptome assembler such as
        <a href=http://trinityrnaseq.sourceforge.net/>Trinity</a>, <a
        href=http://www.ebi.ac.uk/~zerbino/oases/>Oases</a>, or <a
        href=http://www.bcgsc.ca/platform/bioinfo/software/trans-abyss>Trans-ABySS</a>.
        If your organism has a reference genome you can assemble
        transcripts directly from mapped reads using <a
        href="http://cufflinks.cbcb.umd.edu/">Cufflinks</a>. If your genome is already annotated (in
        GTF/GFF), you can generate a multi-FASTA file using the <a href=http://genome.ucsc.edu/>UCSC Genome Browser</a> by uploading your annotation as a track and downloading the sequences under the "Tables" tab.</li>
		  <li>Read alignments to the multi-FASTA file in SAM
        or BAM format.  These can either be stored in a file or
        streamed directly from an aligner. It is important that you
        allow as many multi-mappings as possible. You can also allow
        many mismatches during mapping since eXpress builds an error model to probabalistically assign the reads, although this will increase mapping time.</li>
		  </ol>
		  <h2 id="example">An example</h2>
		  <p>In the following two sub-sections, you will run eXpress on a sample RNA-Seq dataset with simulated reads from UGT3A2 and the HOXC cluster using the human genome build hg18.  Both the transcript sequences (transcripts.fasta) and raw reads (reads_1fastq, reads_2.fastq) can be found in the <tt>&lt;YOUR_EXPRESS_DIR&gt;/sample_data</tt> directory.  For this example to work, you will need to have both <a href=http://bowtie-bio.sourceforge.net>Bowtie</a> and <a href=http://samtools.sf.net>SAMtools</a> installed, but in general any aligner will work and the conversion to BAM is not necessary unless you have insufficient disk space to store the uncompressed SAM.</p>
		  <p>Before you begin, you must prepare your Bowtie index. Since you wish to allow many multi-mappings, it is useful to build the index with a small offrate (in this case 1).  The smaller the offrate, the larger the index and the faster the mapping.  If you have disk space to spare, always use an offrate of 1. Build the index with the following commands.</p>
		  <ol><li><pre class="sc"><code>$ cd &lt;YOUR_EXPRESS_DIR&gt;/sample_data <br />$ bowtie-build --offrate 1 transcripts.fasta transcripts</code></pre></li></ol>
		  <p>This command will populate your directory with several index files that allow Bowtie to more easily align reads to the transcripts.</p>
		  <p>You can now map the reads to the transcript sequences using the following Bowtie command, which outputs in SAM (<code>-S</code>), allows for unlimited multi-mappings (<code>-a</code>), a maximum insert distance of 800 bp between the paired-ends (<code>-X 800</code>), and 3 mismatches (<code>-v 3</code>). The first three options (<code>a,S,X</code>) are highly recommended for best results.  You should also allow for many mismatches, since eXpress models sequencing errors.  Furthermore, you will want to take advantage of multiple processors when mapping large files using the <tt>-p</tt> option.  See the <a href=http://bowtie-bio.sourceforge.net/manual.shtml>Bowtie Manual</a> for more details on various parameters and options.  <br><br>The SAM output from Bowtie is piped into SAMtools in order to compress it to BAM format.  This conversion is optional, but will greatly reduce the size of the alignment file.
		  <ol><li><pre class="sc"><code>$ bowtie -aS -X 800 --offrate 1 -v 3 transcripts -1 reads_1.fastq -2 reads_2.fastq | samtools view -Sb - > hits.bam </code></pre></li></ol></p>
		  <h2 id="fromfile">Input from SAM/BAM file</h2>
		  <p>Once you have aligned your reads to the transcriptome and stored them in a SAM or BAM file, you can run eXpress in default mode with the command:
 		  <ol><li><pre class="sc" margin-left:20px;><code>$ express transcripts.fasta hits.bam</code></pre></li></ol></p>
 		  <p>The default settings will be sufficient for most users, but please see the <a href=manual.html#running>Manual</a> for full descriptions of all available options.</p>
 		  <h2 id="stream">Streaming input from aligner</h2>
 		  <p>If you do not wish to store an intermediate SAM/BAM file, you can pipe the Bowtie output directly into eXpress with the command:</p>
		  <ol><li><pre class="sc"><code>$ bowtie -aS -X 800 --offrate 1 -v 3 transcripts -1 reads_1.fastq -2 reads_2.fastq | express transcripts.fasta </code></pre></li></ol></p>
		  <p>This will give you the exact same output as the
        previous command, while avoiding the need potentially large
        amounts of disk space for storing the mapped reads.</p>
   		  <h2 id="output">Understanding the output</h2>
   		  <p>The output for eXpress is saved in a file called <tt>results.xprs</tt> in an easy-to-parse tab-delimited format.  Since an output directory was not specified, it is simply placed in the working directory (<tt>&lt;YOUR_EXPRESS_DIR&gt;/sample_data</tt>). You can view the output for the run using the command:
   		  <ol><li><pre class="sc"><code>$ more results.xprs</code></pre></li></ol></p>
   		  <p>Your results should look identical to this:</p>
   		  <ol><li><pre class="output"><code>bundle_id	target_id	length	eff_length	tot_counts	uniq_counts	post_count_mode	post_count_var	eff_count_mode	eff_count_var	fpkm	fpkm_conf_low	fpkm_conf_high
1	NM_022658	2288	2111.031696	4881	4881	4881.000000	0.000000	5290.175426	0.000000	231213.960930	224595.002623	237832.919238
2	NM_014212	2037	1860.067174	55	55	55.000000	0.000000	60.231696	0.000000	2956.882460	2159.470561	3754.294359
3	NM_006897	1541	1364.137281	664	664	664.000000	0.000000	750.088729	0.000000	48675.452910	44897.505073	52453.400746
4	NM_001168316	2283	2106.032402	1552	12	516.598635	388.000000	560.007853	455.946043	24529.472300	21673.244046	27385.700554
4	NM_174914	2385	2208.017985	1745	38	930.845799	436.250000	1005.457041	508.987356	42157.527944	38808.437597	45506.618292
4	NR_031764	1853	1676.093181	1243	7	316.555566	310.750000	349.967096	379.809355	18886.513572	15897.888697	21875.138448
5	NM_017410	2396	2219.016430	42	42	42.000000	0.000000	45.349822	0.000000	1892.730465	1308.621162	2476.839768
6	NM_017409	1959	1782.078199	47	47	47.000000	0.000000	51.666083	0.000000	2637.370236	1867.970262	3406.770211
7	NM_173860	849	672.235093	962	962	962.000000	0.000000	1214.958886	0.000000	143104.698051	133876.936763	152332.459340
8	NM_014620	2300	2123.029999	958	77	553.868775	239.500000	600.037768	281.092286	26088.598628	23435.145523	28742.051732
8	NM_153693	2072	1895.062226	481	5	77.654759	95.416638	84.905212	114.066119	4097.741888	2709.327293	5486.156482
8	NR_003084	1640	1463.123288	266	0	21.854792	29.846520	24.496814	37.498989	1493.708156	510.829610	2476.586702
8	NM_153633	1666	1489.119613	762	10	390.173574	190.500000	436.519114	238.443675	26201.627498	22965.195823	29438.059174
8	NM_018953	1612	1435.127246	228	91	210.455804	20.501608	236.393502	25.866471	14664.609325	12546.708897	16782.509753
8	NM_004503	1681	1504.117493	384	37	330.992296	73.025659	369.916614	91.211035	22005.747410	19333.057547	24678.437273
</code></pre></li></ol>
   		  <p>While it may be difficult to read in your
        terminal, opening the file with R or Excel should help you to
        visualize the columns.  An important column is FPKM,
        which reports the estimated abundance in expected
        <b>F</b>ragments <b>P</b>er <b>K</b>ilobase per <b>M</b>illion
        mapped fragments.  Other fields include the mode (post_count_mode) and variance (post_count_var) of the posterior count distribution, as well as the "effective" mode (eff_count_mode) and variance (eff_count_var) after correction for bias.  Transcripts are sorted by their bundle_id, which denotes which multi-mapping group the transcript belongs to and can help determine isoforms and gene families.  More details on the output including full descriptions of all columns can be found in the <a href=manual.html#expr>Manual</a>.</p> 
          		<p>&rarr; <a href="#top">Back to top.</a></p>
         
        <h2 id="diff">Calculating differential expression</h2>
        <p>If you have multiple samples and replicates, you may want to discover if there is a significant change in abundance of any genes or isoforms under different conditions.  Unfortunately, a differential expression tool does not yet exist to take advantage of the full distribution on estimated counts that we output, but we are working on one that will be available soon.  For now, we recommend inputting the rounded effective counts for your samples into a count-based differential expression tool such as <a href='http://www.bioconductor.org/packages/2.6/bioc/html/DEGseq.html'>DEGSeq</a> or <a href='http://www.bioconductor.org/packages/release/bioc/html/edgeR.html'>edgeR</a>.</p>
        <p>&rarr; <a href="#top">Back to top.</a></p>

      </div> 
          
        <div class="corner-content-1col-bottom"></div>                               
	  </div>
      
       <!-- C.2 SUBCONTENT -->
      <div class="subcontent">
      
         <!-- SUBCONTENT CELL -->
        <div class="corner-subcontent-top"></div>                        
        <div class="subcontent-box">
          <h1 class="menu">Outline</h1> 
          <div class="sidemenu1">
          <!-- CONTENT CELL -->                
 		  <ul>
			  <li><a href="#install">Installation </a></li>
				  <ul>
				  <li><a href="#precomp">&rarr;Installing pre-compiled binary</a></li>
				  <li><a href="#src">&rarr;Installing from source</a></li>
				  </ul>
			   <li><a href="#usecase">General Use Case</a></li>
			   <ul>
			   <li><a href="#reqin">&rarr;Required input</a></li>
			   <li><a href="#example">&rarr;An example</a></li>
			   <li><a href="#fromfile">&rarr;Input from SAM/BAM file</a></li>
			   <li><a href="#stream">&rarr;Streaming input from aligner</a></li>
  			   <li><a href="#output">&rarr;Understanding the output</a></li>
  			   <li><a href="#diff">&rarr;Calculating differential expression</a></li>
			   </ul>
		  </ul>          
          </div> 
          </div>
        <div class="corner-subcontent-bottom"></div> 

 	 </div>
    </div>
      
    <!-- D. FOOTER -->      
    <div class="footer">
      <p>Copyright &copy; 2011 Adam Roberts&nbsp;&nbsp;|&nbsp;&nbsp;All Rights Reserved</p>
      <p class="credits">Design by <a href="http://1234.info/" title="Designer Homepage">1234.info</a> | Modified by <a href="http://cs.berkeley.edu/~adarob/">Adam Roberts</a> | <a href="http://validator.w3.org/check?uri=referer" title="Validate XHTML code">XHTML 1.0</a> | <a href="http://jigsaw.w3.org/css-validator/" title="Validate CSS code">CSS 2.0</a></p>
      <br />
      <p>The eXpress project was funded in part by an NSF graduate fellowship to Adam Roberts and NIH grant 1R01HG006129-01</p>
    </div>
    <div class="corner-page-bottom"></div>        
  </div> 
  
</body>
</html>



